tutorials/klusterai-api/effective-prompt-engineering.ipynb

```diff
@@ -0,0 +1,979 @@
+    "\n",
+    "Structure your response clearly with separate sections and explain briefly your reasoning behind each recommendation.\"\"\"\n",
+    "\n",
+    "# Display the prompt for visual reference\n",
+    "display(Markdown(better_prompt_text))\n",
+    "\n",
+    "# Use the variable in the API call\n",
+    "response_better = client.chat.completions.create(\n",
+    "    model=model,\n",
+    "    messages=[{\"role\": \"user\", \"content\": better_prompt_text}],\n",
+    "    temperature=0.5\n",
+    ").choices[0].message.content\n",
+    "print(\"\\nModel response:\\n\" + response_better)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6f3ca7a4",
+   "metadata": {},
+   "source": [
+    "### Analysis\n",
+    "This Better prompt elevates the request beyond the \"Bad\" version by incorporating pricing details and clearer instructions, which yield more focused calculations and recommendations. The prompt also specifies a timeline (January 2025) and outlines key tasks (calculating revenue, identifying product performance, suggesting sales-boosting strategies). As a result, the LLM response provides:\n",
+    "\n",
+    "1. **Structured Revenue Calculations**: By explicitly asking for per-product revenue, it brings a quantitative foundation to the analysis.\n",
+    "2. **Actionable Recommendations**: Requests for pricing and marketing adjustments prompt specific suggestions, rather than broad, generic ideas.\n",
+    "3. **Organized Output**: Clearly defined sections ensure the response is easy to interpret and act upon.\n",
+    "\n",
+    "However, the prompt stops short of requesting profit margins, customer demographics, or broader business context (e.g., marketing channels or inventory challenges) seen in the Best prompt. Consequently, while the \"Better\" prompt moves toward more concrete, data-driven insights, it still lacks the richness and depth needed to produce thorough, strategic recommendations."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a8ceea92",
+   "metadata": {},
+   "source": [
+    "## Best prompt 🟢\n",
+    "\n",
+    "Finally, we arrive at a prompt that provides enough detail for the LLM to respond with meaningful insights on growing the business and increasing profits in this hypothetical scenario. By outlining comprehensive business context, detailed product information, structured analytical requirements, demographic data, profit margins, marketing channels, and specific business challenges, this prompt equips the model to deliver tailored and actionable recommendations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "cb27726c",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/markdown": [
+       "Here's monthly sales data (units sold) from our small retail store for four products during January 2025:\n",
```

Comment:
Similar comment as I had before,  but it's just a lot to have this in the code and then also rendered on the page because it's just duplicating what you see and in a slightly different format. Lots of scrolling to get to what's actually meaningful: the response